src/systems/environmental_sensor_system/EnvironmentalSensorSystem.cc

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#include "EnvironmentalSensorSystem.hh"
+
+#include <gz/plugin/Register.hh>
+
+#include <sdf/Sensor.hh>
+#include <gz/sim/components/CustomSensor.hh>
+#include <gz/sim/components/Sensor.hh>
+#include <gz/sim/components/ParentEntity.hh>
+#include <gz/sim/components/Name.hh>
+#include <gz/sim/components/Environment.hh>
+#include <gz/sim/Util.hh>
+
+#include <gz/sensors/SensorFactory.hh>
+#include <gz/sensors/SensorTypes.hh>
+#include <gz/sensors/Util.hh>
+
+#include <gz/math/Pose3.hh>
+
+#include <chrono>
+#include <unordered_map>
+#include <utility>
+#include <unordered_set>
+#include <memory>
+#include <string>
+
+using namespace gz;
+using namespace gz::sim;
+
+/// Sensor prefix to be used. All envionment_sensors are to be prefixed by
+/// "environment_sensor/" in their gz:type field.
+const char* SENSOR_TYPE_PREFIX =  "environmental_sensor/";
+
+////////////////////////////////////////////////////////////////
+/// \brief Envirtonment Sensor used for looking up environment values in our
+/// CSV file.
+class EnvironmentalSensor : public gz::sensors::Sensor
+{
+  /// \brief Node for communication
+  private: gz::transport::Node node;
+
+  /// \brief Publishes sensor data
+  private: gz::transport::Node::Publisher pub;
+
+  ////////////////////////////////////////////////////////////////
+  /// Documentation inherited
+  public: virtual bool Load(const sdf::Sensor &_sdf) override
+  {
+    auto type = gz::sensors::customType(_sdf);
+    if (type.substr(0, strlen(SENSOR_TYPE_PREFIX)) != SENSOR_TYPE_PREFIX ||
+      type.size() <= strlen(SENSOR_TYPE_PREFIX))
```

Comment:
@arjo129 fyi: there's [`gz::common::StartsWith`](https://github.com/gazebosim/gz-common/blob/da529a8207f20daa8cc33efbe8636a48307dfbdb/include/gz/common/StringUtils.hh#L64) exactly for this kind of checks.